internals/overlord/state/notices.go

```diff
@@ -0,0 +1,395 @@
+// Copyright (c) 2023 Canonical Ltd
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License version 3 as
+// published by the Free Software Foundation.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package state
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"sort"
+	"strconv"
+	"time"
+
+	"github.com/canonical/pebble/internals/logger"
+)
+
+const (
+	MaxNoticeKeyLength = 255
+
+	noticeExpireAfter = 7 * 24 * time.Hour
+)
+
+// Notice represents an aggregated notice. The combination of type and key is unique.
+type Notice struct {
+	// Server-generated unique ID for this notice (a surrogate key). Users
+	// shouldn't rely on this, but this will be a monotonically increasing
+	// number (like change ID).
+	id string
+
+	// The notice's type.
+	noticeType NoticeType
+
+	// The notice key: a string that must be unique for this notice type.
+	// For example, for the "change-update" notice, the key is the change ID.
+	// For a warning, the key would be the warning message.
+	//
+	// This is limited to a maximum of 255 bytes when added (it's an error
+	// to add a notice with a longer key).
+	key string
+
+	// The first time one of these notices (type and key combination) occurred.
+	firstOccurred time.Time
+
+	// The last time one of these notices occurred.
+	lastOccurred time.Time
+
+	// Same as lastOccurred when the notice was last repeated. Only updated
+	// once lastOccurred > lastRepeated + repeatAfter.
+	lastRepeated time.Time
+
+	// The number of times one of these notices has occurred.
+	occurrences int
+
+	// Additional data for the last occurrence of this type and key combination.
+	lastData map[string]string
+
+	// How much time after one of these last occurred should we allow it to repeat.
+	repeatAfter time.Duration
+
+	// How much time since one of these last occurred should we drop the notice.
+	expireAfter time.Duration
+}
+
+func (n *Notice) String() string {
+	return fmt.Sprintf("Notice %s (%s:%s)", n.id, n.noticeType, n.key)
+}
+
+// expired reports whether this notice has expired (relative to the given "now").
+func (n *Notice) expired(now time.Time) bool {
+	return n.lastOccurred.Add(n.expireAfter).Before(now)
+}
+
+// jsonNotice exists so we can control how a Notice is marshalled to JSON. It
+// needs to live in this package (rather than the API) because we save state
+// to disk as JSON.
+type jsonNotice struct {
```

Comment:
Again, modelled after the similar `jsonWarning` struct in `state/warnings.go`, but the reason this is needed is 1) so we can have the fields of `Notice` be private (whereas fields for `encoding/json` have to be exported/capitalized) but still have the fields JSONable, and 2) so that we can (un)marshal `time.Duration` values as "1s" duration strings rather than 1000000000 (integer number of nanoseconds).